[MIG] Port pos_pricelist to 9.0

x620 opened the pull request.

antespi committed:
[IMP] Show unit price when qty is not 1

Conflicts:
	pos_pricelist/__openerp__.py

(cherry-pick commit: e9ac0d5d7ac272666e750fbab6c3a39a9af0357b)

StefanRijnhart committed:
[IMP] Preserve manually modified price in most cases

Conflicts:
	pos_pricelist/static/src/js/models.js

(cherry-pick commit: a10bc113c1b602fa80c725a82045c9b1c2eeaa09)

Commit:
[WIP] App installable now

legalsylvain commented on pos_pricelist/views/point_of_sale_view.xml:
```diff
@@ -10,7 +10,18 @@
+                            <!--</tree>-->
+                        <!--</field>-->
+                    <!--</page>-->
+                <!--</page>-->
```
Hi,

Could your remove this commented useless part of code ?

legalsylvain commented on pos_pricelist/demo/pos_pricelist_demo.yml:
```diff
@@ -62,34 +62,35 @@
         sequence: 1
         min_quantity: 2
-        base: 1
+#        base: 1
```
same here, useless line I guess.

legalsylvain commented:
Hi @x620, 
Thanks for porting this important module CC AdilHoumadi  ! You marked the last commit with 'WIP' tag. is this PR still Work in progress or do you review review ?
some other comment inline.

Other points, You can take a look on the following PR that improve the V8 version :
-  #79 (CC @StefanRijnhart) 
- #71 (CC idem)
- #56 (CC @antespi)

Maybe you could git cherry-pick all / some commits done against V8, since V9 branches has been created.

Thanks again and kind regards.

Commit:
Add contributors

Commit:
Add contributors

pedrobaeza committed:
[FIX] pos_pricelist: FP taxes with multi-company. Fixes #84

The object account.fiscal.position.tax doesn't have multi-company rules,
which leads to error in environments with multi-company and these records,
because it finally accesses to the parent fiscal position, which on contrary
has access rules, provoking an access error.

This commit added the field company_id to the model, and add the corresponding
access rule.

Conflicts:
	pos_pricelist/__openerp__.py

Commit:
Corrected error: TypeError: Cannot read property 'field' of undefined (and many others before it)
Error occurred: TypeError: Cannot read property 'models' of undefined (in progress)

yelizariev commented on pos_pricelist/static/src/js/models.js:
```diff
@@ -15,24 +15,217 @@
+    var round_pr = models.round_precision;
+    var round_di = models.round_decimals;
+
+	PosDB = PosDB.extend({
```
Move this block to db.js

Commit:
Recovery db.js

Commit:
Corrected error: TypeError: Cannot read property 'models' of undefined

Commit:
Corrected error: TypeError: round_pr is not a function

x620 commented:
During running the application an errors does not appear. What to look for and what to consider in order to be sure that the application works correctly?

Commit:
Corrected error: AttributeError: 'float' object has no attribute 'decimal_places'
Corrected error: KeyError: 'total'

Commit:
Corrected error: TypeError: _amount_line_tax() got an unexpected keyword argument 'context'

flotho commented:
Hi @x620 @yelizariev ,

I've made some additional test in your db.
We have a pricelist : http://3143544-83-bcbce2.runbot1.odoo-community.org/web#id=3&view_type=form&model=product.pricelist&menu_id=113&action=122
And agrolait is affected to the pricelist : http://3143544-83-bcbce2.runbot1.odoo-community.org/web#id=7&view_type=form&model=res.partner&menu_id=68&action=54
Yet when selling products of the pricelist for agrolait in the POS the price si not discounted.
Could it come from the fact that the list price item is based on template instead of variant?
I also made some tests :
- First line is on variant
- second on template
- Rest as a global fucntion with discount

None of those rules are applied

Commit:
Recovery tooltips (in progress)

Commit:
Recovery tooltips: corrected compute_price

flotho commented:
Hi @x620 , 

Do you want us to make additionnal tests?

x620 commented:
Hi flotho. 
Yet only tooltips.

flotho commented:
Ok, we'll wait, did you got the test use cases we made ?

regards

x620 commented:
Yes. Thanks

flotho commented:
Thanks to you!

Commit:
Check discount values in pricelist cards and remove outdated fields

Commit:
Corrected calculate discount and remove outdated code

x620 commented:
Flotho, I made a few corrections. Can you to make additional tests?

flotho commented:
Hi @x620 ,

We made additionnal tests.
We closed the session and restart new ones, change the pricelist, use fixed price, formula, etc... without success.

Our workflow of testing :
- Start a brand new session
- Create a new order
- Use agrolait and set the customer on each new transaction
- Choose one of the products in the pricelist (those with fixed price, percentage, etc...)

=> the price used in the order line is not the correct one

flotho commented:
@x620 does the links specified here https://github.com/OCA/pos/pull/83#issuecomment-195376973 still in the correct DB?
Maybe we are not working on the correct one.

flotho commented:
test updated here : http://3143822-9-0-0b7bdb.runbot2.odoo-community.org/web#id=4&view_type=form&model=product.pricelist&action=228
Customer agrolait
No changes in the pos.order.line prices : http://3143822-9-0-0b7bdb.runbot2.odoo-community.org/web?#id=1&view_type=form&model=pos.order&menu_id=557&action=783
Sorry ;-)

Commit:
Added percent_price and fixed_price method define discount

flotho commented:
Hi @x620 don't hesitate to tell us when to test.

regards

x620 commented:
I think it is too early. I prefer to move in small steps. I have not yet all completed.

Commit:
Add author in __openerp__.py

Commit:
Few corrections

x620 commented:
Hi flotho. I corrected display price in order card and few other corrections. I think that it is possible to start additional test.

Commit:
Add images and license keys

yelizariev commented on pos_pricelist/__openerp__.py:
```diff
@@ -20,9 +20,12 @@
+    'author': "Adil Houmadi @Taktik,Odoo Community Association (OCA), \
+    IT-Projects LLC, Pavel Romanchenko",
     'summary': 'Pricelist for Point of sale',
+    'license': 'LGPL-3',
```
change license to AGPL-3

Commit:
Corrected license value

Commit:
Corrected outdated parameters

x620 commented:
Somebody knows that does mean this error in Travis CI test:

2016-03-17 13:21:09,180 23857 CRITICAL openerp_test openerp.service.server: Failed to initialize database `openerp_test`.
Traceback (most recent call last):
File "/home/travis/OCB-9.0/openerp/service/server.py", line 885, in preload_registries
registry = RegistryManager.new(dbname, update_module=update_module)
File "/home/travis/OCB-9.0/openerp/modules/registry.py", line 386, in new
openerp.modules.load_modules(registry._db, force_demo, status, update_module)
File "/home/travis/OCB-9.0/openerp/modules/loading.py", line 419, in load_modules
cr.execute("SELECT name FROM ir_module_module WHERE state='installed' and name in %s", (tuple(mods),))
File "/home/travis/OCB-9.0/openerp/sql_db.py", line 141, in wrapper
return f(self, _args, *_kwargs)
File "/home/travis/OCB-9.0/openerp/sql_db.py", line 220, in execute
res = self._obj.execute(query, params)
ProgrammingError: syntax error at or near ")"
LINE 1: ...FROM ir_module_module WHERE state='installed' and name in ()

How to relations the error to the pos_pricelist module?

pedrobaeza commented:
It was a problem we had yesterday with OCB, but it's solved. I have relaunch the build.

flotho commented:
Hi, Same thing here : http://3143940-83-a86483.runbot1.odoo-community.org/web?debug=#id=3&view_type=form&model=product.pricelist&menu_id=113&action=122
Am I forgetting something in my conf? The pricelist is attached to agrolait and is not working

flotho commented:
in the console :

```
error: Some modules could not be started 
Failed modules:          ["pos_pricelist.widgets"] 
Debug:                   Object {pos_pricelist.widgets: Object}
```

yelizariev commented on pos_pricelist/static/src/js/widgets.js:
```diff
@@ -66,7 +72,7 @@ function pos_pricelist_widgets(instance, module) {
     });
 
-    module.PosBaseWidget.include({
+    screens.PosBaseWidget.include({
```
PosBaseWidget can be reached as following:

```
 var PosBaseWidget = require('point_of_sale.BaseWidget')
```

flotho commented:
@yelizariev , will @x620 push this modification soon?

Commit:
Correction errors display discounts related to the pricelist customers

Commit:
Refactoring

x620 commented:
Hi, flotho. Pricelist work with Global, Product Category and Product Variant. Product doesn't work, but it seems it doesn't work in 8 too.

flotho commented:
Hi, my project manager validate your work, perfect thanks

flotho commented:
Hi have some issue with fiscal positions when loading the POS
It is said that I haven't got access right.
Once your module is deactivate it works,
I have only some customers that have fiscal position.
What could be the issue?
Regards

yelizariev commented:
Sounds like user doesn't have access to **account.fiscal.position.tax**
https://github.com/x620/pos/blob/pos-9.0-pos_pricelist/pos_pricelist/static/src/js/models.js#L651

legalsylvain commented:
hi, 
I don't think if it's related, but an access right bug of Odoo core on 'account.fiscal.position.tax' model is fixed in this V 8.0 OCA PR https://github.com/OCA/pos/pull/85
CC : @pedrobaeza 
Regards.

flotho commented:
Could it be related to multi company environment?

flotho commented:
Obviously it is,

I tested with 2 multi company rules for account.fiscal.position.tax and account.fiscal.position.account and it works now,
@x620 could you add those rules and fields  based on the model of @pedrobaeza 

Thanks

pedrobaeza commented:
I has a similar problem and make this PR: https://github.com/OCA/pos/pull/85

flotho commented:
@x620, @yelizariev is there anything special with pos_cache that could prevent the module to work?

regards

yelizariev commented:
@flotho I think updates from #85 should be applied when that request is merged

flotho commented:
@yelizariev , the #85 is for v8 ;-)

flotho commented:
@yelizariev , @x620,

what is your position regarding my latest comment? 

Thanks

yelizariev commented:
@flotho those updates have to be reviewed and accepted for 8.0 first, then we can apply it for 9.0

flotho commented:
ok, I understand the process.
I'll implement it form my side as it solve the issue

flotho commented:
@yelizariev,

I have another problem, locally I have no issue in .js yet the pricelist don't seems to apply. I have combinations of products / category etc...
What is the correct workflow? Close the sessions? reinstall/refresh the pricelist....
I have about 45000 products and about 60000 customers so loading is long long...

yelizariev commented:
@flotho try check that js is loaded in broswer

flotho commented:
@yelizariev, .js is not loaded even in the Back Office, I'll dig in and let you know

flotho commented:
@yelizariev could it be conflicts with your pos-addons repository ?

yelizariev commented:
@flotho there could be conflicts with any modules

flotho commented:
Hi @x620 , @yelizariev 

Does the field display_price_with_taxes still usefull as Odoo now have field called iface_tax_included ?

flotho commented:
Hi @x620 , @yelizariev ,

I have strange issues while loading :
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14089619/149fac64-f536-11e5-9138-f15cfb1b70c7.jpeg)
And more.

But it looks like there is a conflict with pos_disable_payment and multi session.

Could it come from the pos_cache ?

Any advice would be appreciated,

regards

flotho commented:
Additionnal tests here :
- Desinstaling multi session and disable payment make the module work

Regarding the functions, I worked with this pricelist :
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14090298/bd915f5e-f539-11e5-869c-6151d951e9b7.jpeg)
- The pricelist don't work with fixed price => How can I help you to debug this one
- The pricelist works with % reductions

As you can see here :
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14090323/e07ce59c-f539-11e5-90d1-78fcfe3729b9.jpeg)

test florent is correctly used with the reduction, but not the "florent bestiole" that should be a fixed price

yelizariev commented:
@flotho 

> Does the field display_price_with_taxes still usefull as Odoo now have field called iface_tax_included ?

probably, we can use  iface_tax_included.

yelizariev commented:
@flotho https://github.com/OCA/pos/pull/83#issuecomment-202580863
do you reproduce this issue on runbot, that is without additional modules?

flotho commented:
Hi  @yelizariev , 

Sorry for having previously said that it was fully functionnal on runbot. 
We have the same issue here regarding https://github.com/OCA/pos/pull/83#issuecomment-202580863 :
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14099303/823050a0-f584-11e5-91ba-4678fec85dd8.jpeg)
here is the result :
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14099316/a9bdc9ea-f584-11e5-9299-dcc0fcb3b0f2.jpeg)

Some kind of sequence might not be used in your code as the default global seems to apply correctly with % discount.

yelizariev commented:
Do you mean that module doesn't apply sequence of pricelist rules?

flotho commented:
I'm wondering if it is not the problem. As you could see on my first test, fixed price are not applied but discount and global are, even on the products that have fixed price. As they are the first in the list, I suspect this kind of behaviour with sequence.
Can't it be an indication?

regards

yelizariev commented:
@flotho 
It seems that 8.0 has the same behavior: price calculation is stopped after first found rule
https://github.com/OCA/pos/blob/8.0/pos_pricelist/static/src/js/models.js#L530

legalsylvain commented:
https://github.com/OCA/pos/pull/85 (for v8) has been merged. could you cherry pick that commit in this PR ? https://github.com/OCA/pos/pull/85/commits/b9221e27a3bd88b1331cf3c573b50d831ff2f269

kind regards.

yelizariev commented:
@x620 
could please make cherry-pick:

```
git fetch upstream
git cherry-pick b9221e27a3bd88b1331cf3c573b50d831ff2f269
git push

```

flotho commented:
@yelizariev ,

I'm not sure to understand your answer regarding the sequence. I understand that once a rule is found for a product, it's unnecessary to go further, yet as you can see in my previous screenshots, "2L Evian" with fixed price is my first rule regarding the new way of sorting rules in odoo v9. So for this product, this rule should have been preferred to the default one which is global.
As you know, pricelist structure has changed, for example, here is the content of the pricelist I first displayed :

```
SELECT pricelist_id, sequence, fixed_price, create_date, price_discount, applied_on
       percent_price, date_start, company_id, product_tmpl_id,  
       price_min_margin, compute_price, base, write_date, categ_id, 
       price_surcharge, price_round, product_id, base_pricelist_id
  FROM product_pricelist_item
  WHERE pricelist_id=8;

Pricelist   seq fixed_price appliedon       
8           5                           3_global        formula
8           5   456                 0_product_variant   0.000   fixed
8           5   0                   0_product_variant   0.000   percentage

```

As you can see Odoo make its own sort and always finished and display list price with global at the end, and the sequence is now used in a second step.
My opinion is that the method used to sort https://github.com/OCA/pos/blob/9.0/pos_pricelist/static/src/js/db.js#L128 has to be rewrite a little so that the type of rule is preferred as a first key for sorting.

What do you think of this ?

yelizariev commented:
@flotho ok, thanks for explanation

@x620 could you please add sorting on applied_on field so that 3_global will beat  the end

Commit:
Added sort rules by applied_on field

Commit:
Added applied_on field

flotho commented:
@yelizariev , @x260, a great thanks for this.

regards

flotho commented:
@yelizariev , @x620 , I have an additionnal question regarding categories.
If you have 2 categories like :
A - Main category and B-child category which is a child of A
How do you deal with products in Categ B if you set up a pricelist rule on category A? Do you browse all products belonging to the tree of the category or directly attached on the category?
Sorting the rules by category with the lower level first could be a way of implementing this .

Thanks for your feedback!

yelizariev commented:
@flotho  
for products in category B system checks is there rule for category B and if there is, such rule is applied, if not system checks the same for parent of category B and so on
https://github.com/OCA/pos/blob/8.0/pos_pricelist/static/src/js/models.js#L452

flotho commented:
@yelizariev , thanks

Great News!

flotho commented:
Hi all of you, personnaly I validate that the functionnal part is working.
FMO, it could be merged.

regards

flotho commented:
Hi @x620 , @yelizariev , 
I have a strange issue that I didin't succeed to debug. On my localhost it works perfectly, on my preprod I have such a result :
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14386760/64841bc0-fda6-11e5-86ae-bdd2804ff041.jpeg)

for those rules :
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14386772/7ed0cd20-fda6-11e5-8a34-a36900993547.jpeg)

The behaviour is as the prices in rules weren't not used. I'm up to date and working with this one : https://github.com/MindAndGo/pos

Could you lead me for debugging? 

regards

flotho commented:
@yelizariev @x620 ,

I have a question. I made a test with a single pricelist in which there are some rules for a specific variant regarding to the quantity. If I select a customer, the correct prices are displayed, if I don't choos the pricelist, the rules are recognised but not the prices.
It looks like a bug to me, shouldn't the price be updated with each rules even if no customer is used?

regards

flotho commented:
Regarding the point of the rules that are not correctly applied, I think I've found what's going on. If you add a price to a product, it's added by default with a rule on template.
If you change it later and want to set it up on a variant, the information on template still exists :
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14420775/2264276a-ffce-11e5-8ab7-a52656185262.jpeg)
As you can see, some rules keeps together variant_id AND template id. So my opinion is to not exclude based on template rules IF the variant one is selected and also if applied on is set to variant.
Could it be reviewed?

yelizariev commented:
@flotho as I understand, you asking to change current behavior https://github.com/OCA/pos/pull/83#issuecomment-202774766

flotho commented:
@yelizariev , @x620 ,

regarding the default behaviour on pricelist when no partner is chosen, could it come from the switch https://github.com/OCA/pos/pull/83/files#diff-747f62729cafe771e948f5cfeb9aff46R456 and the different cases used?

yelizariev commented:
@flotho this code could be the reason.

```
                if (!partner || rule.pricelist_id[1] != partner.property_product_pricelist[1]) {
                        continue
            }
```

@x620 could you explain why do you do such check? Doesn't mean, that pricelist is not applied if there is no customer choosen?
https://github.com/x620/pos/commit/95173e245cb9ce5b2acda616586c322e162093c7

flotho commented:
@yelizariev , thanks for your answer 

> @flotho as I understand, you asking to change current behavior #83 (comment)

Not really, my point is not to replace this behaviour but to make it stronger. I explained. 
As you see on my latest screenshot, the pricelist rule is applied_on variant YET odoo (not your code) let the information about the template in the rule.
So my point is that somewhere in your code you are giving the priority to a control on the field template_id instead of the applied_on field .
So the previous behaviour should not been replaced but enhanced so that the logic could be respected even if Odoo don't update all the fields correctly.

Tell me if I'm miss something in my explanations

regards

yelizariev commented:
Do you mean to give rules on Template more priorotiy that on Variant? I don't think that it's a good idea.

```
    'applied_on': fields.selection([('3_global', 'Global'),('2_product_category', ' Product Category'), ('1_product', 'Product'), ('0_product_variant', 'Product Variant')], string="Apply On", required=True,
```

flotho commented:
@yelizariev , not at all, the applied on is ok.
My point is that when a rule keeps the template info (bug from odoo ?) even if it's specified as applied_on variant, the price computation failedand keep the default price.
That's the point of https://github.com/OCA/pos/pull/83#issuecomment-208223616 .

And it's mostly related to this part of code : https://github.com/OCA/pos/pull/83/files#diff-747f62729cafe771e948f5cfeb9aff46R417

flotho commented:
Hi @yelizariev, to summarize we still have :
- 1 bug related to https://github.com/OCA/pos/pull/83#issuecomment-208252262
- 1 possible bug when there are rules based on product (1_product) seems to not be applied!

flotho commented:
@x620 , @yelizariev ,

Any feedback on those 2 little bugs? regards

yelizariev commented:
@flotho  updates are coming soon

flotho commented:
@yelizariev , thanks for the feedback,

Commit:
[FIX] bug when find item

Commit:
[FIX] bug when find item

flotho commented:
@x620 , @yelizariev ,

Do you want me to test now?
regards

yelizariev commented:
@flotho yes, try now

flotho commented:
Hi  @yelizariev , @x620 ,

i've made additionnal tests, here are the results :
- When no customer is selected the default pricelist is applied with correct quantities and amount => OK
- When rules are defined for template => KO:
  ![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14519486/a037e19c-021f-11e6-9e0b-a3fb1da860e0.jpeg)
  it still not working:
  ![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14519504/bded3778-021f-11e6-8d42-bc4ebb12619d.jpeg)

tested here : http://3145222-83-f7812c.runbot2.odoo-community.org/web#id=2&view_type=form&model=product.pricelist&menu_id=113&action=122

Hope it could help

yelizariev commented on pos_pricelist/static/src/js/models.js:
```diff
@@ -410,18 +416,20 @@ function pos_pricelist_models(instance, module) {
-                    || categ_ids.indexOf(item.categ_id[0]) !== -1) &&
-                    (item.price_version_id[0] === version.id)) {
+                if (
+                    (item.product_tmpl_id === false || item.product_tmpl_id[0] === product.id)
```
I believe it must be  like this:

```
  (item.product_tmpl_id === false || item.product_tmpl_id[0] === product.product_tmpl_id)
```

Commit:
[FIX] check product_tmpl_id when collect rules (db.js) and items (models.js)

yelizariev commented:
@flotho try again

flotho commented:
Hi @yelizariev ,

It's better, we're almost done.
Here is y conf :
Default price list
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14536453/b7630fa6-0272-11e6-860a-9d4c8d6ce9f6.jpeg)
Specific price lis associated to Agrolait
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14536476/d1a35aa6-0272-11e6-864c-cbfca10f8cc9.jpeg)

Here are the results of my tests  :

1 - Without any customer and a  => almost OK, why does the evian template display 2 lines?
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14536715/0f721948-0274-11e6-8242-2f9f231ef3a0.jpeg)

2 - Start a new order with customer rules displayed seems to be uncomplete (it behaves correctly regarding to the quantity) :
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14536751/498f3fd4-0274-11e6-9fba-bf0321ce1950.jpeg)

In the end, the function seems to be Ok but it looks like there is a regression on the displayed popups.

x620 commented:
Hi, @flotho 
In both cases tooltips also work and in 8.
![odoo8_a](https://cloud.githubusercontent.com/assets/17245696/14560111/fa1bc9b4-0326-11e6-9d48-32456f756387.png)

![odoo8_b](https://cloud.githubusercontent.com/assets/17245696/14560127/138d005c-0327-11e6-9849-bddde70c4e2c.png)

But in the second case, it can be easily eliminated by adding a row in the pricelist: All products 1x 0% discount. Then we get:

![odoo8_c](https://cloud.githubusercontent.com/assets/17245696/14560134/1dd17354-0327-11e6-8a65-fcda1b2f0443.png)

flotho commented:
Ok, it just looks weird to me but still functionnal.
So we could say that it's Ok

flotho committed:
Add return on method for API call convenience

flotho commented:
Hi @yelizariev , @x620 ,

I have a strange conflict when  I use this module with others. The product.product model is not found while loaded so that some fields are missing in the model.
It looks like using initialize.apply mess with another part. What is the correct way to push fileds for product.product in other modules so that there won't be conflicts with yours?

flotho committed:
Force order of product fields loading

flotho committed:
Merge branch 'pos-9.0-pos_pricelist' into 9.0

flotho commented:
Hi @yelizariev , @x620 ,

i'm not sure that it is te most proper way to deal with my issue but I change a little bit your syntax https://github.com/MindAndGo/pos/commit/9c79f80ccdecaf6d3df9cc8bd6267a9be2f787c0 and it works for me. The categ_id are always loaded.
I'm sure there is a conflicts with other modules loading the product.product model but I can't find what's the proper way to deal with that so, I use this patch.
Your feedback would be appreciated.

regards

flotho commented:
Hi @yelizariev , @x620 ,

Last bug really touchy I think. I tested on the latest runbot.
I have the main pricelist with quantities :
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14814832/51240350-0ba8-11e6-856d-b6796d582bfc.jpeg)

Then another one pricelist that depends on the first one :
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14814858/6e81b97e-0ba8-11e6-9d43-b41d34f70684.jpeg)

Once you select your customer, your module succeed to identify that there are many lines but the price is not correctly used
![odoo_v10](https://cloud.githubusercontent.com/assets/1043935/14814874/83a7278a-0ba8-11e6-9e4a-07c5564f788c.jpeg)

tested there : http://3146099-83-7c28e6.runbot1.odoo-community.org/web#id=2&view_type=form&model=product.pricelist&menu_id=113&action=122

yelizariev commented:
> It looks like using initialize.apply mess with another part. What is the correct way to push fileds for product.product in other modules so that there won't be conflicts with yours?

It should be inheritance instead of overwriting, because overwriting reject all previous inheritances
https://odoo-development.readthedocs.org/en/latest/dev/pos/add_field.html

flotho commented:
Hi @yelizariev ,

> It should be inheritance instead of overwriting, because overwriting reject all previous inheritances
> https://odoo-development.readthedocs.org/en/latest/dev/pos/add_field.html

I tried this one and have some problems after... I'll regive it a try

Commit:
[FIX] fix handle where rule.base == 'pricelist'

x620 commented:
Hi @flotho 
You can now test.

StefanRijnhart commented:
It seems you kept all the changes that are intended to support fiscal positions in the POS. This was a rather 'hidden' functionality of this module, and it was not completely functional. As the point_of_sale module in Odoo 9.0 has native support for fiscal positions, you should weed out all these changes.

StefanRijnhart commented:
If you are nice, you migrate the cross table that is used for pos.order.line's tax_ids in the 8.0 version of this module to the cross table that is used in 9.0's point_of_sale for the same field in a migration script.

flotho commented:
Hi @x620 ,

i'll test.
Could you also deal with this part : https://github.com/OCA/pos/pull/83#issuecomment-215033132 and integrate it in the next commit?

@yelizariev , what about the proposal of @StefanRijnhart , could it be included in my latest financial support?

yelizariev commented:
@flotho 
pos_pricelist does make inheritance and not overwriting https://github.com/x620/pos/blob/pos-9.0-pos_pricelist/pos_pricelist/static/src/js/models.js#L703-L711 
I meant that other modules can do overwriting that broke system

yelizariev commented:
@StefanRijnhart is it correct, that all files in models/ folder are related to fiscal position and can be removed?

flotho commented:
@yelizariev 
I'm no more sure to understand because you mentionned https://odoo-development.readthedocs.io/en/latest/dev/pos/add_field.html . Does that means that we have to .call or to .apply?

Anyway, thanks for your support

flotho committed:
Merge branch 'pos-9.0-pos_pricelist' of https://github.com/x620/pos into pos-9.0-pos_pricelist

flotho committed:
Merge branch 'pos-9.0-pos_pricelist' into 9.0

yelizariev commented:
@flotho it's not a big difference between .call and .apply

You can find an example of wrong way to add fields here - https://github.com/yelizariev/pos-addons/blob/8.0/tg_pos_enhanced/static/src/js/tg_pos.js#L17
-- it comletly overwrites load_server_data function

StefanRijnhart commented:
@yelizariev looks like it, except for the definition of the display_price_with_taxes field, which is another side feature of this module.

flotho commented:
@x620 , @yelizariev 

From my point of view, it works !

phoenixsampras commented:
Awesome work!! please keep it on!! this is so necessary for supermarkets...

woepwoep commented:
pls connect with me re supermarkets
i am in the food cooperative industry, helping to build a movement here in
the Netherlands.
contact details below
thx, Ron
web: www.voedsel.coop
email: info@voedsel.coop
phone: +31 629 584 582

---

http://about.me/woepwoep

2016-05-20 5:01 GMT+02:00 phoenixsampras notifications@github.com:

> Awesome work!! please keep it on!! this is so necessary for supermarkets...
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/pos/pull/83#issuecomment-220506064

flotho commented:
@yelizariev ,

I have a remark regarding performances. I think it becomes useless to add supplierinfo in the POS.
Actually, the default behaviour of pricelist don't use supplierinfo anymore as way of computing prices.
What is your point of view?

regards

flotho committed:
[IMP]Prevent useless supplierinfo loading

flotho committed:
[IMP]Deal with dates in pricelist items

yelizariev commented:
I suggest to try merge this PR to new branch, say 9.0-pos_pricelist. This could simplify contributing by other developers. If someone find issue and find a way to fix, one will be able to create new PR instead of asking @x620 and wait when he update his PR.

See discussion on twitter: https://twitter.com/jgrandguillaume/status/735738547118825475

The same approach could be applied to other PR with `needs review` status

Commit:
Merge pull request #1 from MindAndGo/date_on_items

Date on items

gaelTorrecillas commented:
hello,

@yelizariev in your models.js, the function models.PosModel.extend's initialize returns nothing, why?
When I have to overload this function, I put "return _super_posmodel.initialize.call(this, session, attributes);" 
I understand that you used .apply but we don't return _super_posmodel.initialize.apply(this, arguments) ?

Commit:
[FIX] initialize() method with ...initialize.apply(this, arguments); must have return

Commit:
[FIX] Moved ...initialize.apply(this, arguments); to its original place

gaelTorrecillas commented:
thanks x620 for your job.

I have other question, in
https://github.com/x620/pos/blob/pos-9.0-pos_pricelist/pos_pricelist/static/src/js/models.js#L262

the function "can_be_merged_with", why the inherit? because, if the products don't have the same id, the core returns false, and if the products' price aren't the same, the core returns false.
do we want save the inherit or if we don't save it, the point of sale will not have the problem?

dcoopennet commented:
@x620 Is this PR going to be merged ? We need this PORT for one of our customers.
I can review it if you want.

rafaelbn commented:
Hi @x620 , thanks a lot for this PR, thanks all for revisions. This PR is amazing! 😄 

Sorry for the late.

Please @x620 could you rebase to check again what's going on with travis?

Also it could be useful is you check this FIX: https://github.com/OCA/pos/pull/100 made by @llacroix in v8

Thanks again!

rafaelbn commented:
Hi again, as commented by @yelizariev in https://github.com/OCA/pos/pull/83#issuecomment-223221180 after rebase we will check travis and runbot but we can merge if all is green. cc @x620 @legalsylvain

yelizariev commented:
@rafaelbn 
Pavel (@x620) is out of business and would not make updates. So, it's probably better to create new PR with rebasing and other updates. Can you do it yourself?

rafaelbn commented:
I cannot myself, but I could ask in my team. In this moment we don't need this but I wish we will. If someone is interested in POS in v9 then could do it. Let wait sometime then.

rafaelbn commented:
Hi all! Is anyone interested in this PR which is a migration from v8 to v9 of `pos_pricelist`. If not I will close this PR. Thanks!

pedrobaeza commented:
As a side note, Odoo v11 include this functionality in core.

pedrobaeza commented:
Closing this in favor of #186
